<a href="https://colab.research.google.com/github/dinujakr/Statistical-Learning-e20190/blob/main/Assignments/data_analysis_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Foundations of Statistical Inference & Hypothesis Testing**

Objective of this assignment is to rigorously evaluate the stationarity of an operational baseline and master the mathematical machinery required to control false-alarm risks in continuous diagnostic layers.

## **Part A: Theoretical Fundamentals (Maximum Likelihood & Decision Space)**



1. Consider the sequence of i.i.d. random variables  
$$\mathbf{X}_1,\mathbf{X}_2,\cdots, \mathbf{X}_n\in \mathbb{R}^m$$
with $\boldsymbol{\mu}=\mathbb{E}(\mathbf{X}_i)$ and $\boldsymbol{\Sigma}=\mathrm{Var}(\mathbf{X}_i)$. The unobservable multivariate population parameters $(\boldsymbol{\mu}, \boldsymbol{\Sigma})$ are estimated with their corresponding unbiased empirical sample estimators $(\widehat{\boldsymbol{\mu}}_n, \mathbf{S})$, where
$$\widehat{\boldsymbol{\mu}}_n = \frac{1}{n}\sum_{i=1}^n \mathbf{X}_{i},$$
and
$$\mathbf{S} = \frac{1}{n-1} \sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T .$$

* Formally derive why the raw Maximum Likelihood Estimator for the covariance matrix, $\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}$, is inherently **biased** in finite samples, by showing that:

$$\mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \frac{n-1}{n}\boldsymbol{\Sigma}$$


* Mathematically demonstrate how applying Bessel's correction yields the unique, unbiased estimator $\mathbf{S}$.


2. Under the null hypothesis $H_0$ that a structural asset remains completely healthy (the status quo), a chosen test statistic tracks a deterministic parametric distribution.
* Define a **Type I Error ($\alpha$)** and a **Type II Error ($\beta$)** explicitly within the context of continuous structural health monitoring.
* If an engineer sets an ultra-conservative significance threshold (e.g., $\alpha = 0.0001$) to prevent false alarms from shutting down an industrial plant, explain the mathematical consequence this decision exerts on the statistical power of the diagnostic layer. How does this modify the geometric volume of your "healthy operation" confidence ellipsoid?

#### 1. Bias of the MLE Covariance Matrix and Bessel's Correction

Given

$$
\widehat{\boldsymbol{\mu}}_n=\frac{1}{n}\sum_{i=1}^{n}\mathbf{X}_i,
$$

the Maximum Likelihood Estimator (MLE) of the covariance matrix is

$$
\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}
=
\frac{1}{n}
\sum_{i=1}^{n}
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)^T.
$$

Using the identity

$$
\sum_{i=1}^{n}
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)^T
=
\sum_{i=1}^{n}
(\mathbf{X}_i-\boldsymbol{\mu})
(\mathbf{X}_i-\boldsymbol{\mu})^T
-
n(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})^T,
$$

we obtain

$$
\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}
=
\frac{1}{n}
\sum_{i=1}^{n}
(\mathbf{X}_i-\boldsymbol{\mu})
(\mathbf{X}_i-\boldsymbol{\mu})^T
-
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})^T.
$$

Taking expectations,

$$
\mathbb{E}
\left[
\frac{1}{n}
\sum_{i=1}^{n}
(\mathbf{X}_i-\boldsymbol{\mu})
(\mathbf{X}_i-\boldsymbol{\mu})^T
\right]
=
\boldsymbol{\Sigma},
$$

and since

$$
\mathrm{Var}(\widehat{\boldsymbol{\mu}}_n)
=
\frac{\boldsymbol{\Sigma}}{n},
$$

it follows that

$$
\mathbb{E}
\left[
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})^T
\right]
=
\frac{\boldsymbol{\Sigma}}{n}.
$$

Therefore,

$$
\mathbb{E}
[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}]
=
\boldsymbol{\Sigma}
-
\frac{\boldsymbol{\Sigma}}{n}
=
\frac{n-1}{n}\boldsymbol{\Sigma}.
$$

Hence,

$$
\boxed{
\mathbb{E}
[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}]
=
\frac{n-1}{n}\boldsymbol{\Sigma}
}
$$

showing that the MLE covariance estimator is biased downward.

To remove this bias, apply Bessel's correction:

$$
\mathbf{S}
=
\frac{n}{n-1}
\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}
=
\frac{1}{n-1}
\sum_{i=1}^{n}
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)^T.
$$

Taking expectations,

$$
\mathbb{E}[\mathbf{S}]
=
\frac{n}{n-1}
\mathbb{E}
[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}]
=
\frac{n}{n-1}
\left(
\frac{n-1}{n}
\boldsymbol{\Sigma}
\right)
=
\boldsymbol{\Sigma}.
$$

Therefore,

$$
\boxed{
\mathbb{E}[\mathbf{S}]
=
\boldsymbol{\Sigma}
}
$$

and

$$
\boxed{
\mathbf{S}
=
\frac{1}{n-1}
\sum_{i=1}^{n}
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)^T
}
$$

is the unbiased sample covariance estimator.

#### 2. Type I Error, Type II Error, and Effect of an Ultra-Conservative Significance Level

Consider the hypotheses

$$
H_0:\text{The structure is healthy}
$$

and

$$
H_1:\text{The structure is damaged}.
$$

### Type I Error ($\alpha$)

A Type I error occurs when the monitoring system incorrectly declares damage even though the structure is healthy.

$$
\boxed{
\alpha
=
P(\text{Reject }H_0 \mid H_0 \text{ is true})
}
$$

In structural health monitoring, this corresponds to a false alarm that may trigger unnecessary inspections or plant shutdowns.

### Type II Error ($\beta$)

A Type II error occurs when the monitoring system fails to detect actual damage.

$$
\boxed{
\beta
=
P(\text{Fail to reject }H_0 \mid H_1 \text{ is true})
}
$$

In practice, this means that structural degradation remains undetected.

### Effect of Choosing an Extremely Small Significance Level

Suppose the engineer chooses

$$
\alpha = 0.0001.
$$

To achieve such a small false-alarm probability, the acceptance region under $H_0$ must be enlarged.

For a multivariate monitoring statistic based on the Mahalanobis distance,

$$
(\mathbf{x}-\boldsymbol{\mu})^T
\boldsymbol{\Sigma}^{-1}
(\mathbf{x}-\boldsymbol{\mu})
\le c_\alpha,
$$

where

$$
c_\alpha=\chi^2_{m,\,1-\alpha}.
$$

As $\alpha$ decreases,

$$
c_\alpha
=
\chi^2_{m,\,1-\alpha}
\quad \text{increases},
$$

and therefore the confidence ellipsoid expands.

The volume of the healthy-operation ellipsoid is proportional to

$$
V
\propto
c_\alpha^{m/2}
\sqrt{\det(\boldsymbol{\Sigma})}.
$$

Hence,

$$
\boxed{
\alpha \downarrow
\;\Longrightarrow\;
c_\alpha \uparrow
\;\Longrightarrow\;
V \uparrow
}
$$

meaning that a larger region of the feature space is classified as healthy.

### Effect on Statistical Power

The statistical power of the monitoring system is

$$
\boxed{
\text{Power}
=
1-\beta
}
$$

which is the probability of correctly detecting damage.

Because the healthy acceptance region becomes larger when $\alpha$ is reduced, more damaged states fall inside that region. Consequently,

$$
\boxed{
\alpha \downarrow
\;\Longrightarrow\;
\beta \uparrow
\;\Longrightarrow\;
(1-\beta)\downarrow
}
$$

Thus, an ultra-conservative significance level greatly reduces false alarms but also decreases the power of the diagnostic system, making true structural damage more difficult to detect.

## **Part B: Theoretical Extension (Asymptotic Distributions & Slutsky's Theorem)**

The Lindeberg-Lévy Multivariate Central Limit Theorem establishes that as the sample snapshot size approaches infinity ($n \to \infty$), the standardized difference between the empirical mean and population mean converges in distribution to a Multivariate Gaussian:

$$\sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \xrightarrow{d} \mathscr{N}(\mathbf{0}, \boldsymbol{\Sigma})$$



However, the true population covariance matrix $\boldsymbol{\Sigma}$ is a latent, unobservable parameter.
* State **Slutsky’s Theorem** rigorously.
* Prove how Slutsky's Theorem justifies substituting the empirical sample covariance matrix $\mathbf{S}$ for the latent population matrix $\boldsymbol{\Sigma}$ as $n \to \infty$, thereby operationalizing the practical parametric modeling distribution:

$$\widehat{\boldsymbol{\mu}}_n \sim \mathscr{N}\left({\boldsymbol{\mu}}, \frac{1}{n}\mathbf{S}\right)$$







#####By the Lindeberg-Lévy Multivariate Central Limit Theorem,

$$
\sqrt{n}
\left(
\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu}
\right)
\xrightarrow{d}
\mathscr{N}
\left(
\mathbf{0},
\boldsymbol{\Sigma}
\right).
$$

Since the population covariance matrix $\boldsymbol{\Sigma}$ is unknown, it must be replaced by the sample covariance matrix $\mathbf{S}$.

## Slutsky's Theorem

Let

$$
X_n \xrightarrow{d} X
$$

and

$$
Y_n \xrightarrow{p} c,
$$

where $c$ is a constant.

Then:

1.

$$
X_n + Y_n
\xrightarrow{d}
X + c,
$$

2.

$$
X_n Y_n
\xrightarrow{d}
cX,
$$

3.

$$
\frac{X_n}{Y_n}
\xrightarrow{d}
\frac{X}{c},
\qquad c\neq 0.
$$

More generally, if a function $g(\cdot,\cdot)$ is continuous,

$$
g(X_n,Y_n)
\xrightarrow{d}
g(X,c).
$$

---

## Consistency of the Sample Covariance Matrix

The sample covariance matrix

$$
\mathbf{S}
=
\frac{1}{n-1}
\sum_{i=1}^{n}
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)
(\mathbf{X}_i-\widehat{\boldsymbol{\mu}}_n)^T
$$

is a consistent estimator of $\boldsymbol{\Sigma}$.

Therefore,

$$
\mathbf{S}
\xrightarrow{p}
\boldsymbol{\Sigma}.
$$

Consequently,

$$
\mathbf{S}^{-1/2}
\xrightarrow{p}
\boldsymbol{\Sigma}^{-1/2},
$$

by the Continuous Mapping Theorem.

---

## Application of Slutsky's Theorem

From the Multivariate Central Limit Theorem,

$$
\sqrt{n}
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})
\xrightarrow{d}
\mathscr{N}
(\mathbf{0},\boldsymbol{\Sigma}).
$$

Premultiplying by $\mathbf{S}^{-1/2}$ gives

$$
\mathbf{S}^{-1/2}
\sqrt{n}
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu}).
$$

Since

$$
\mathbf{S}^{-1/2}
\xrightarrow{p}
\boldsymbol{\Sigma}^{-1/2},
$$

Slutsky's Theorem implies

$$
\mathbf{S}^{-1/2}
\sqrt{n}
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})
\xrightarrow{d}
\boldsymbol{\Sigma}^{-1/2}
\,\mathscr{N}
(\mathbf{0},\boldsymbol{\Sigma}).
$$

Using the covariance transformation rule,

$$
\boldsymbol{\Sigma}^{-1/2}
\,\mathscr{N}
(\mathbf{0},\boldsymbol{\Sigma})
=
\mathscr{N}
(\mathbf{0},\mathbf{I}_m).
$$

Hence,

$$
\boxed{
\mathbf{S}^{-1/2}
\sqrt{n}
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})
\xrightarrow{d}
\mathscr{N}
(\mathbf{0},\mathbf{I}_m)
}
$$

which is the same asymptotic limit obtained if the true covariance matrix were known.

---

## Practical Parametric Modeling Distribution

Because

$$
\mathbf{S}
\xrightarrow{p}
\boldsymbol{\Sigma},
$$

the unknown covariance matrix in the asymptotic distribution can be replaced by its consistent estimator without affecting the limiting distribution.

Therefore,

$$
\sqrt{n}
(\widehat{\boldsymbol{\mu}}_n-\boldsymbol{\mu})
\overset{\text{approx}}{\sim}
\mathscr{N}
(\mathbf{0},\mathbf{S}),
$$

and dividing the covariance by $n$ yields

$$
\boxed{
\widehat{\boldsymbol{\mu}}_n
\overset{\text{approx}}{\sim}
\mathscr{N}
\left(
\boldsymbol{\mu},
\frac{1}{n}\mathbf{S}
\right)
}
$$

for sufficiently large sample sizes.

Thus, Slutsky's Theorem provides the theoretical justification for replacing the latent population covariance matrix $\boldsymbol{\Sigma}$ with the observable sample covariance matrix $\mathbf{S}$ when constructing practical asymptotic Gaussian models and confidence regions.

## **Part C: Numerical Verification**

***Scenario:** You are provided an industrial dataset containing $n = 5,000$ snapshots across $m = 3$ sensor channels measuring physical strain. To verify the weak i.i.d. assumption before initializing a global PCA or Factor Analysis layer, you must explicitly test **Condition 1: Constant Joint Mean (First Moment Homogeneity)** across the time index.*

**Your Task:**
Write a clean, self-contained Python function within Google Colab to partition the dataset into $g = 5$ consecutive, non-overlapping blocks (chunks) of size $n_j = 1,000$. Implement a Multivariate Analysis of Variance (MANOVA) tracking framework from scratch (or leveraging `statsmodels`) to calculate **Wilks' Lambda ($\Lambda$)**.

$$\Lambda = \frac{|\mathbf{W}|}{|\mathbf{B} + \mathbf{W}|}$$

Transform your calculated $\Lambda$ using **Bartlett’s $\chi^2$ approximation**:


$$\chi^2_{\text{calc}} = -\left(n - 1 - \frac{m + g}{2}\right) \ln \Lambda$$

Execute your code using the synthetic data generator provided below. Report your exact calculated Wilks' $\Lambda$, the Bartlett $\chi^2$ test statistic, and its corresponding $p$-value under $\chi^2 \sim \chi^2_{m(g-1)}$. Based on an alpha threshold of $\alpha = 0.05$, state your final diagnostic conclusion: Has the baseline center shifted over time, or is the first moment homogeneous?



In [5]:
import numpy as np
import pandas as pd
import scipy.stats as stats

np.random.seed(42)
n_samples = 5000
n_features = 3

base_data = np.random.multivariate_normal(
    mean=[0.5, -0.2, 1.1],
    cov=[[0.09, 0.02, 0.01],
         [0.02, 0.06, 0.03],
         [0.01, 0.03, 0.05]],
    size=n_samples
)

base_data[4000:, 0] += 0.015
base_data[4000:, 2] -= 0.010

df_strain = pd.DataFrame(base_data, columns=['Strain_Ch1', 'Strain_Ch2', 'Strain_Ch3'])

In [6]:
def verify_first_moment_homogeneity(df: pd.DataFrame, g_chunks: int = 5) -> dict:

    X = df.values
    n, m = X.shape

    if n % g_chunks != 0:
        raise ValueError("n must be divisible by g_chunks")

    block_size = n // g_chunks
    global_mean = np.mean(X, axis=0)

    W = np.zeros((m, m))
    B = np.zeros((m, m))

    for j in range(g_chunks):
        chunk = X[j*block_size:(j+1)*block_size]
        chunk_mean = np.mean(chunk, axis=0)

        W += (chunk - chunk_mean).T @ (chunk - chunk_mean)

        diff = (chunk_mean - global_mean).reshape(m, 1)
        B += block_size * (diff @ diff.T)

    Lambda = np.linalg.det(W) / np.linalg.det(W + B)

    chi2_calc = -(n - 1 - (m + g_chunks)/2) * np.log(Lambda)

    df_chi = m * (g_chunks - 1)

    p_value = 1 - stats.chi2.cdf(chi2_calc, df_chi)

    decision = (
        "Reject H0: mean shift detected"
        if p_value < 0.05
        else
        "Fail to reject H0: first-moment homogeneity holds"
    )

    return {
        "Wilks_Lambda": Lambda,
        "Chi_square": chi2_calc,
        "df": df_chi,
        "p_value": p_value,
        "decision": decision
    }

In [7]:
results = verify_first_moment_homogeneity(df_strain, g_chunks=5)
print(results)

{'Wilks_Lambda': np.float64(0.9904519796929591), 'Chi_square': np.float64(47.9215049961488), 'df': 12, 'p_value': np.float64(3.2255259508895406e-06), 'decision': 'Reject H0: mean shift detected'}


### MANOVA Test for First-Moment Homogeneity (Constant Joint Mean)

We partition the dataset into:

$$
g = 5 \quad \text{blocks}, \quad n_j = 1000, \quad n = 5000, \quad m = 3.
$$

Let $\bar{\mathbf{x}}_j$ be the mean of block $j$, and $\bar{\mathbf{x}}$ the global mean.

## Within-group scatter matrix

$$
\mathbf{W}
=
\sum_{j=1}^{g}
\sum_{i=1}^{n_j}
(\mathbf{x}_{ij}-\bar{\mathbf{x}}_j)
(\mathbf{x}_{ij}-\bar{\mathbf{x}}_j)^T
$$

## Between-group scatter matrix

$$
\mathbf{B}
=
\sum_{j=1}^{g}
n_j
(\bar{\mathbf{x}}_j-\bar{\mathbf{x}})
(\bar{\mathbf{x}}_j-\bar{\mathbf{x}})^T
$$

## Wilks' Lambda

$$
\Lambda
=
\frac{|\mathbf{W}|}{|\mathbf{W}+\mathbf{B}|}
$$

## Bartlett's Chi-square approximation

$$
\chi^2_{\text{calc}}
=
-\left(n - 1 - \frac{m + g}{2}\right)\ln(\Lambda)
$$

with degrees of freedom:

$$
df = m(g-1)
$$

## Decision rule

- Reject $H_0$ if $p < 0.05$
- Otherwise, fail to reject $H_0$

Where:

$$
H_0: \boldsymbol{\mu}_1 = \boldsymbol{\mu}_2 = \cdots = \boldsymbol{\mu}_5
$$


# **Geometric Subspace Optimization via Principal Component Analysis (PCA)**

The objective of this assignment is to master the rigorous linear algebra mechanics underpinning coordinate transformations, verify total variance conservation laws, and design operational monitoring thresholds using the Hotelling's $T^2$ and $Q$ (SPE) statistics.

## **Part A: Theoretical Fundamentals (Coordinate Projections & Orthogonality)**

1. Let $\widetilde{\mathbf{X}}_i = (\mathbf{X}_i - \boldsymbol{\mu}) \in \mathbb{R}^m$ represent a zero-mean, centered baseline feature vector with a symmetric, positive-semidefinite population covariance matrix $\boldsymbol{\Sigma} = \mathbb{E}[\widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T]$. We apply the spectral decomposition $\boldsymbol{\Sigma} = \mathbf{P} \mathbf{\Lambda} \mathbf{P}^T$, and define the orthogonal projection mapping:

$$\mathbf{Z}_i = \mathbf{P}^T \widetilde{\mathbf{X}}_i$$


* Prove step-by-step that the theoretical covariance matrix of the transformed vector, $\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T]$, simplifies exactly to the strictly diagonal matrix $\mathbf{\Lambda}$.
* Explain the precise geometric and statistical meaning of this diagonal result regarding the cross-covariances between distinct coordinates $Z_{i,j}$ and $Z_{i,k}$ (for $j \neq k$).


2. The total variance of a multivariate system can be measured in its raw coordinate space or its transformed principal component space.
* Utilizing the algebraic properties of the matrix trace ($\text{tr}$), specifically its invariance under cyclic permutations, formally prove that the total system variance is preserved during the mapping from raw space to the de-correlated principal space:

$$\text{tr}(\boldsymbol{\Sigma}) = \text{tr}(\mathbf{\Lambda}) = \sum_{j=1}^m \lambda_j$$


* If we partition the empirical matrix into an explained subspace ($\widehat{\mathbf{P}}_k \in \mathbb{R}^{m \times k}$) and an unexplained residual subspace ($\widehat{\mathbf{P}}_{m-k} \in \mathbb{R}^{m \times (m-k)}$), state the rigorous mathematical formulations for the **Cumulative Explained Variance Ratio $\Phi(k)$** and the **Residual Unexplained Variance Ratio $\Psi(k)$**.


3. In structural health monitoring (SHM), we truncate our coordinate system to retain only the first $k$ components ($k < m$). The centered raw vector is partitioned as $(\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n) = \widehat{\mathbf{P}}_k \mathbf{z}_{i,k} + \widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k}$.
* Define the reconstructed vector $\widehat{\mathbf{x}}_i$ and the residual error vector $\mathbf{e}_i$.
* Prove via the Pythagorean identity that the total squared Euclidean norm of the centered observation vector cleanly decomposes into independent subspace energies:

$$\|\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n\|^2 = \|\mathbf{z}_{i,k}\|^2 + \|\mathbf{z}_{i,m-k}\|^2$$


* Contrast the diagnostic functions of Hotelling's $T^2$ statistic and the residual $Q$ statistic (SPE). If a structure encounters an environmental load anomaly (e.g., extreme operational wind loads) versus an internal structural fracture (e.g., a fatigue crack changing the structural load pathways), match each event type to the statistic ($T^2$ or $Q$) most sensitive to it and justify your choices.


### Part A: Coordinate Projections & Orthogonality in PCA

Let

$$
\widetilde{\mathbf{X}}_i = (\mathbf{X}_i - \boldsymbol{\mu}) \in \mathbb{R}^m,
\qquad
\boldsymbol{\Sigma} = \mathbb{E}[\widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T].
$$

Assume spectral decomposition:

$$
\boldsymbol{\Sigma} = \mathbf{P}\mathbf{\Lambda}\mathbf{P}^T,
$$

where $\mathbf{P}$ is orthonormal ($\mathbf{P}^T\mathbf{P}=\mathbf{I}$), and

$$
\mathbf{\Lambda} = \mathrm{diag}(\lambda_1,\dots,\lambda_m).
$$

Define the PCA-transformed vector:

$$
\mathbf{Z}_i = \mathbf{P}^T \widetilde{\mathbf{X}}_i.
$$

---

## 1. Covariance in PCA coordinates

We compute:

$$
\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T]
=
\mathbb{E}[\mathbf{P}^T \widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T \mathbf{P}]
$$

Pulling out constants:

$$
=
\mathbf{P}^T \mathbb{E}[\widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T] \mathbf{P}
$$

Substitute covariance:

$$
=
\mathbf{P}^T \boldsymbol{\Sigma} \mathbf{P}
$$

Now substitute spectral decomposition:

$$
=
\mathbf{P}^T (\mathbf{P}\mathbf{\Lambda}\mathbf{P}^T)\mathbf{P}
$$

Using orthogonality:

$$
\mathbf{P}^T\mathbf{P}=\mathbf{I}
$$

we get:

$$
\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T]
=
\mathbf{\Lambda}.
$$

### Result:

$$
\boxed{
\mathrm{Cov}(\mathbf{Z}_i)=\mathbf{\Lambda}
}
$$

---

## Geometric and statistical meaning

Since $\mathbf{\Lambda}$ is diagonal:

$$
\mathrm{Cov}(Z_{i,j}, Z_{i,k}) = 0 \quad \text{for } j \neq k.
$$

### Interpretation:

- PCA rotates the coordinate system into an orthogonal eigenbasis of $\boldsymbol{\Sigma}$.
- Each coordinate $Z_{i,j}$ lies along an eigenvector direction.
- Zero off-diagonal terms imply **complete decorrelation** between principal components.
- Geometrically: the covariance ellipsoid becomes axis-aligned in PCA space.
- Statistically: each principal component carries independent variance energy (second-order decorrelation).

---

## 2. Total variance preservation

Total variance in original space:

$$
\mathrm{Var}_{\text{total}} = \mathrm{tr}(\boldsymbol{\Sigma})
$$

Using spectral decomposition:

$$
\mathrm{tr}(\boldsymbol{\Sigma})
=
\mathrm{tr}(\mathbf{P}\mathbf{\Lambda}\mathbf{P}^T)
$$

Using cyclic invariance of trace:

$$
\mathrm{tr}(\mathbf{P}\mathbf{\Lambda}\mathbf{P}^T)
=
\mathrm{tr}(\mathbf{\Lambda}\mathbf{P}^T\mathbf{P})
$$

Since $\mathbf{P}^T\mathbf{P}=\mathbf{I}$:

$$
=
\mathrm{tr}(\mathbf{\Lambda})
$$

Thus:

$$
\boxed{
\mathrm{tr}(\boldsymbol{\Sigma}) = \mathrm{tr}(\mathbf{\Lambda}) = \sum_{j=1}^m \lambda_j
}
$$

---

## Explained and residual variance ratios

Let $\widehat{\mathbf{P}}_k$ contain the first $k$ eigenvectors.

### Cumulative Explained Variance Ratio

$$
\Phi(k)
=
\frac{\sum_{j=1}^{k} \lambda_j}
{\sum_{j=1}^{m} \lambda_j}
$$

### Residual Unexplained Variance Ratio

$$
\Psi(k)
=
\frac{\sum_{j=k+1}^{m} \lambda_j}
{\sum_{j=1}^{m} \lambda_j}
=
1 - \Phi(k)
$$

---

## 3. Reconstruction, residuals, and energy decomposition

The centered vector is decomposed as:

$$
\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n
=
\widehat{\mathbf{P}}_k \mathbf{z}_{i,k}
+
\widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k}.
$$

### Reconstruction

$$
\widehat{\mathbf{x}}_i
=
\widehat{\boldsymbol{\mu}}_n + \widehat{\mathbf{P}}_k \mathbf{z}_{i,k}
$$

### Residual vector

$$
\mathbf{e}_i
=
\mathbf{x}_i - \widehat{\mathbf{x}}_i
=
\widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k}
$$

---

## Pythagorean decomposition

Because PCA bases are orthonormal:

$$
\widehat{\mathbf{P}}_k^T \widehat{\mathbf{P}}_{m-k} = \mathbf{0}
$$

Thus energy splits orthogonally:

$$
\|\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n\|^2
=
\|\mathbf{z}_{i,k}\|^2 + \|\mathbf{z}_{i,m-k}\|^2
$$

---

## SHM interpretation: T² vs Q statistic

### Hotelling’s $T^2$

$$
T^2 = \mathbf{z}_{i,k}^T \mathbf{\Lambda}_k^{-1} \mathbf{z}_{i,k}
$$

- Measures deviation **inside PCA model subspace**
- Sensitive to **changes in dominant correlation structure**
- Captures:
  - sensor bias shifts
  - environmental loading changes
  - operational condition variation

---

### Squared Prediction Error (Q statistic / SPE)

$$
Q = \|\mathbf{z}_{i,m-k}\|^2
$$

- Measures energy in **residual subspace**
- Sensitive to **unmodeled structure**
- Captures:
  - sensor faults
  - structural damage (cracks, stiffness loss)
  - changes not explained by learned PCA space

---

## Event matching

### 1. Environmental load anomaly (e.g., wind gusts, operational overload)
→ **Detected by: $T^2$**

Reason:
- stays within learned correlation structure
- shifts projection inside PCA space
- increases modeled variance, not residual energy

---

### 2. Internal structural fracture (e.g., fatigue crack)
→ **Detected by: $Q$ (SPE)**

Reason:
- changes physical coupling between sensors
- violates learned covariance structure
- produces new directions outside PCA subspace

---

## Final summary

| Event Type | Sensitive Statistic | Reason |
|------------|--------------------|--------|
| Environmental load variation | $T^2$ | stays in PCA subspace |
| Structural damage / fracture | $Q$ | creates residual energy outside model |

---
```


## **Part B: Numerical Verification**



***Scenario:** You are building a real-time diagnostics layer for a structural component monitored by $m = 4$ sensors. You have a baseline archive of $n = 3,000$ nominal snapshots. To establish operational boundaries, you must write an optimization pipeline that analyzes how Hotelling’s $T^2$ and the residual $Q$ statistic respond to different truncation boundaries $k \in \{1, 2, 3\}$.*

**Your Task:**
Complete the Python class method below inside Google Colab. The function must:

1. Standardize and center the incoming realizations.
2. Decompose the unbiased sample covariance matrix $\mathbf{S}$ using `np.linalg.eigh` and sort components in descending order ($\widehat{\lambda}_1 \ge \dots \ge \widehat{\lambda}_m$).
3. Loop through every possible subspace cutoff value $k \in \{1, 2, 3\}$, calculating the sample vector array for Hotelling’s $T^2$ ($T^2_i = \sum_{j=1}^k \frac{z_{i,j}^2}{\widehat{\lambda}_j}$) and the $Q$ statistic ($Q_i = \sum_{j=k+1}^m z_{i,j}^2$).
4. Compute the expected value (empirical mean) of $T^2$ and $Q$ across all samples for each $k$.
5. Return these metrics to verify your theoretical derivations in Part A.



In [8]:

import numpy as np

class PCA_SHM_Diagnostics:
    def __init__(self, X: np.ndarray):
        """
        X: (n, m) baseline data matrix
        """
        self.X = X
        self.n, self.m = X.shape

        # Step 1: center data
        self.mu = np.mean(X, axis=0)
        self.Xc = X - self.mu

        # Step 2: covariance matrix (unbiased)
        self.S = np.cov(self.Xc, rowvar=False, bias=False)

        # Eigen-decomposition
        eigvals, eigvecs = np.linalg.eigh(self.S)

        # Sort in descending order
        idx = np.argsort(eigvals)[::-1]
        self.lambdas = eigvals[idx]
        self.P = eigvecs[:, idx]

        # Projected scores
        self.Z = self.Xc @ self.P

    def evaluate_subspaces(self, k_values=(1, 2, 3)):
        """
        Computes expected T^2 and Q statistics for each truncation k.
        """

        results = {}

        for k in k_values:

            Z_k = self.Z[:, :k]
            Z_res = self.Z[:, k:]

            lambdas_k = self.lambdas[:k]

            # Avoid division issues
            inv_lambdas_k = np.where(lambdas_k > 1e-12, 1.0 / lambdas_k, 0.0)

            # Hotelling T^2 for each sample
            T2 = np.sum(Z_k**2 * inv_lambdas_k, axis=1)

            # Q statistic (SPE)
            Q = np.sum(Z_res**2, axis=1)

            results[k] = {
                "E[T^2]": np.mean(T2),
                "E[Q]": np.mean(Q),
                "T2_vector": T2,
                "Q_vector": Q
            }

        return results


# -------------------------
# Example usage (assumes X is provided)
# -------------------------
# X shape: (3000, 4)

np.random.seed(42)
X = np.random.multivariate_normal(
    mean=[0, 0, 0, 0],
    cov=np.array([
        [1.0, 0.4, 0.2, 0.1],
        [0.4, 1.2, 0.3, 0.2],
        [0.2, 0.3, 1.1, 0.25],
        [0.1, 0.2, 0.25, 0.9]
    ]),
    size=3000
)

model = PCA_SHM_Diagnostics(X)
results = model.evaluate_subspaces(k_values=(1, 2, 3))

for k, v in results.items():
    print(f"\n=== k = {k} ===")
    print("E[T^2]:", v["E[T^2]"])
    print("E[Q]:", v["E[Q]"])



=== k = 1 ===
E[T^2]: 0.9996666666666679
E[Q]: 2.372449308141401

=== k = 2 ===
E[T^2]: 1.9993333333333358
E[Q]: 1.4210282374109602

=== k = 3 ===
E[T^2]: 2.9990000000000023
E[Q]: 0.6817602750648416


# **Latent Subspace Decomposition via Factor Analysis (FA)**

The objective is to master the generative model equations of Factor Analysis, derive factor scores from scratch using conditional expectations (Thomson's method), and numerically isolate sensor malfunctions from true structural damage by tracking communalities ($h^2$) and uniquenesses ($\varphi^2$).

## **Part A: Theoretical Fundamentals (Generative Model & Commonalities)**

1. Consider your structured data setup where $\mathbf{Z}_i \in \mathbb{R}^m$ is a normalized, zero-mean baseline feature vector whose population covariance matrix represents the universal correlation matrix $\mathbf{R} = \mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T]$. The Factor Analysis model posits the linear generative structure:
$$\mathbf{Z}_i = \boldsymbol{\Lambda} \mathbf{F}_i + \boldsymbol{\epsilon}_i$$
where $\mathbf{F}_i \in \mathbb{R}^k$ are the orthogonal common factors ($\mathbb{E}[\mathbf{F}_i\mathbf{F}_i^T] = \mathbf{I}_{k \times k}$) and $\boldsymbol{\epsilon}_i \in \mathbb{R}^m$ is the diagonalized measurement noise matrix ($\mathbb{E}[\boldsymbol{\epsilon}_i\boldsymbol{\epsilon}_i^T] = \boldsymbol{\Psi} = \text{diag}(\varphi_1^2, \dots, \varphi_m^2)$).
* Utilizing the constraint that the latent factors and errors are independent ($\mathbb{E}[\boldsymbol{\epsilon}_i \mathbf{F}_i^T] = \mathbf{0}$), prove the **Fundamental Equation of Factor Analysis**:

$$\mathbf{R} = \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$$


* Focus on the $j$-th diagonal entry of this decomposition. Explicitly define the **Communality ($h_j^2$)** and **Uniqueness ($\varphi_j^2$)** parameters, and explain their physical interpretations in an asset diagnostics system.


2. In Factor Analysis, we often apply an orthogonal coordinate rotation to the factor loading matrix (such as the **Varimax** rotation criterion) to improve model interpretability.
* Contrast how a raw principal component loading vector differs conceptually from a Varimax-rotated factor loading vector.
* Explain why orthogonal rotations can alter the individual columns of the loading matrix $\boldsymbol{\Lambda}$ without changing the calculated communality ($h_j^2$) of any sensor or modifying the global covariance approximation $\mathbf{R} \approx \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$.



1. Because the common factors $\mathbf{F}_i$ are unobservable latent parameters, they cannot be directly extracted via geometric projections like PCA. Instead, we must treat them as hidden random vectors and calculate their realization using **Thomson’s Regression Method**, which relies on the conditional expectation $\mathbf{f}_i = \mathbb{E}[\mathbf{F}_i | \mathbf{z}_i]$.
* Construct the joint partitioned vector $\mathbf{Y}_i = [\mathbf{Z}_i^T, \mathbf{F}_i^T]^T$ and prove that its joint population covariance matrix expands exactly to:

$$\boldsymbol{\Sigma}_{YY} = \begin{bmatrix} \mathbf{R} & \boldsymbol{\Lambda} \\ \boldsymbol{\Lambda}^T & \mathbf{I}_{k \times k} \end{bmatrix}$$


* Under the standard assumption of a Multivariate Normal distribution, the conditional mean of a partitioned system is governed by the projection identity $\mathbb{E}[\mathbf{X}_2 | \mathbf{x}_1] = \boldsymbol{\mu}_2 + \boldsymbol{\Sigma}_{21} \boldsymbol{\Sigma}_{11}^{-1} (\mathbf{x}_1 - \boldsymbol{\mu}_1)$. Using this theorem, derive the final score matrix equation for Thomson’s estimator:

$$\mathbf{f}_i = \boldsymbol{\Lambda}^T \mathbf{R}^{-1} \mathbf{z}_i$$



### Part A: Factor Analysis, Covariance Structure, and Thomson Estimation

Let

$$
\mathbf{Z}_i = \boldsymbol{\Lambda}\mathbf{F}_i + \boldsymbol{\epsilon}_i
$$

with assumptions:

$$
\mathbb{E}[\mathbf{F}_i]=\mathbf{0}, \quad \mathbb{E}[\mathbf{F}_i\mathbf{F}_i^T]=\mathbf{I}_k
$$

$$
\mathbb{E}[\boldsymbol{\epsilon}_i]=\mathbf{0}, \quad \mathbb{E}[\boldsymbol{\epsilon}_i\boldsymbol{\epsilon}_i^T]=\boldsymbol{\Psi}
$$

$$
\mathbb{E}[\mathbf{F}_i \boldsymbol{\epsilon}_i^T]=\mathbf{0}
$$

---

## 1. Fundamental Equation of Factor Analysis

Compute covariance:

$$
\mathbf{R} = \mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T]
$$

Substitute model:

$$
\mathbf{R}
=
\mathbb{E}[(\boldsymbol{\Lambda}\mathbf{F}_i + \boldsymbol{\epsilon}_i)
(\boldsymbol{\Lambda}\mathbf{F}_i + \boldsymbol{\epsilon}_i)^T]
$$

Expand:

$$
=
\boldsymbol{\Lambda}\mathbb{E}[\mathbf{F}_i\mathbf{F}_i^T]\boldsymbol{\Lambda}^T
+
\mathbb{E}[\boldsymbol{\epsilon}_i\boldsymbol{\epsilon}_i^T]
+
\boldsymbol{\Lambda}\mathbb{E}[\mathbf{F}_i\boldsymbol{\epsilon}_i^T]
+
\mathbb{E}[\boldsymbol{\epsilon}_i\mathbf{F}_i^T]\boldsymbol{\Lambda}^T
$$

Using assumptions (cross terms vanish):

$$
\mathbf{R}
=
\boldsymbol{\Lambda}\mathbf{I}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}
$$

Hence:

$$
\boxed{
\mathbf{R} = \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}
}
$$

---

### Communality and Uniqueness

For variable $j$, diagonal entry:

$$
R_{jj} = \sum_{r=1}^k \lambda_{jr}^2 + \varphi_j^2
$$

Define:

### Communality

$$
\boxed{
h_j^2 = \sum_{r=1}^k \lambda_{jr}^2
}
$$

### Uniqueness

$$
\boxed{
\varphi_j^2 = \text{specific variance of variable } j
}
$$

---

### Interpretation (SHM context)

- $h_j^2$: variance of sensor $j$ explained by shared structural dynamics (global modes, load paths)
- $\varphi_j^2$: sensor-specific noise or localized effects (measurement noise, local defects, calibration error)

High communality ⇒ sensor strongly coupled to global structural behavior  
High uniqueness ⇒ sensor dominated by local or noisy effects

---

## 2. Raw vs Varimax-rotated loadings

Let:

- Raw loadings: $\boldsymbol{\Lambda}$
- Orthogonal rotation: $\mathbf{T}$ such that $\mathbf{T}^T\mathbf{T}=\mathbf{I}$

Rotated loadings:

$$
\boldsymbol{\Lambda}^* = \boldsymbol{\Lambda}\mathbf{T}
$$

### Conceptual difference

- PCA loadings: maximize variance sequentially, mathematically optimal but often dense and hard to interpret
- Varimax loadings: orthogonally rotated to maximize variance concentration per factor → improves sparsity and interpretability

---

### Why communality is invariant

Communality for variable $j$:

$$
h_j^2 = \sum_{r=1}^k \lambda_{jr}^2
$$

After rotation:

$$
h_j^{*2}
=
\sum_{r=1}^k (\boldsymbol{\Lambda}\mathbf{T})_{jr}^2
$$

Since $\mathbf{T}$ is orthogonal:

$$
\boldsymbol{\Lambda}\mathbf{T}(\boldsymbol{\Lambda}\mathbf{T})^T
=
\boldsymbol{\Lambda}\mathbf{T}\mathbf{T}^T\boldsymbol{\Lambda}^T
=
\boldsymbol{\Lambda}\boldsymbol{\Lambda}^T
$$

Thus:

- reconstructed covariance unchanged
- diagonal elements unchanged ⇒ communality unchanged

---

### Key result

Orthogonal rotations:
- redistribute variance across factors
- do NOT change total explained variance per variable
- do NOT change covariance reconstruction:

$$
\mathbf{R} \approx \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}
$$

remains invariant

---

## 3. Thomson regression (factor score estimation)

### Joint vector

$$
\mathbf{Y}_i =
\begin{bmatrix}
\mathbf{Z}_i \\
\mathbf{F}_i
\end{bmatrix}
$$

Compute covariance:

- $\mathrm{Cov}(\mathbf{Z}_i)=\mathbf{R}$
- $\mathrm{Cov}(\mathbf{F}_i)=\mathbf{I}$
- $\mathrm{Cov}(\mathbf{Z}_i,\mathbf{F}_i)=\boldsymbol{\Lambda}$

Thus:

$$
\boxed{
\boldsymbol{\Sigma}_{YY}
=
\begin{bmatrix}
\mathbf{R} & \boldsymbol{\Lambda} \\
\boldsymbol{\Lambda}^T & \mathbf{I}
\end{bmatrix}
}
$$

---

### Conditional expectation (Gaussian identity)

$$
\mathbb{E}[\mathbf{F}_i|\mathbf{Z}_i=\mathbf{z}_i]
=
\mathbf{0}
+
\boldsymbol{\Lambda}^T \mathbf{R}^{-1}\mathbf{z}_i
$$

Thus Thomson estimator:

$$
\boxed{
\mathbf{f}_i = \boldsymbol{\Lambda}^T \mathbf{R}^{-1}\mathbf{z}_i
}
$$

---

## Covariance of transformed vector

From PCA rotation:

$$
\mathbf{Z}_i = \mathbf{P}^T \widetilde{\mathbf{X}}_i
$$

Compute:

$$
\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T]
=
\mathbf{P}^T \boldsymbol{\Sigma} \mathbf{P}
=
\mathbf{\Lambda}
$$

---

### Interpretation

For $j \neq k$:

$$
\mathrm{Cov}(Z_{i,j}, Z_{i,k}) = 0
$$

Meaning:
- principal components are second-order uncorrelated
- covariance ellipsoid becomes axis-aligned
- each coordinate represents independent variance direction

---

## 4. Total variance preservation

Using trace:

$$
\mathrm{tr}(\boldsymbol{\Sigma})
=
\mathrm{tr}(\mathbf{P}\mathbf{\Lambda}\mathbf{P}^T)
$$

Cyclic invariance:

$$
=
\mathrm{tr}(\mathbf{\Lambda}\mathbf{P}^T\mathbf{P})
=
\mathrm{tr}(\mathbf{\Lambda})
$$

Thus:

$$
\boxed{
\mathrm{tr}(\boldsymbol{\Sigma}) = \sum_{j=1}^m \lambda_j
}
$$

---

### Variance ratios

$$
\boxed{
\Phi(k)
=
\frac{\sum_{j=1}^k \lambda_j}{\sum_{j=1}^m \lambda_j}
}
$$

$$
\boxed{
\Psi(k)
=
\frac{\sum_{j=k+1}^m \lambda_j}{\sum_{j=1}^m \lambda_j}
= 1-\Phi(k)
}
$$

---

## 5. Reconstruction and SHM decomposition

### Reconstruction

$$
\widehat{\mathbf{x}}_i
=
\widehat{\boldsymbol{\mu}}_n + \widehat{\mathbf{P}}_k \mathbf{z}_{i,k}
$$

### Residual

$$
\mathbf{e}_i
=
\widehat{\mathbf{P}}_{m-k}\mathbf{z}_{i,m-k}
$$

---

## Pythagorean decomposition

Orthogonality:

$$
\widehat{\mathbf{P}}_k^T\widehat{\mathbf{P}}_{m-k}=\mathbf{0}
$$

Thus:

$$
\|\mathbf{x}_i-\widehat{\boldsymbol{\mu}}_n\|^2
=
\|\mathbf{z}_{i,k}\|^2 + \|\mathbf{z}_{i,m-k}\|^2
$$

---

## SHM interpretation: T² vs Q

### Hotelling’s T²

$$
T^2 = \mathbf{z}_{i,k}^T \boldsymbol{\Lambda}_k^{-1}\mathbf{z}_{i,k}
$$

- monitors variation inside PCA model space
- sensitive to:
  - environmental loading changes
  - operational condition shifts

---

### Q statistic (SPE)

$$
Q = \|\mathbf{z}_{i,m-k}\|^2
$$

- monitors deviation outside PCA subspace
- sensitive to:
  - structural damage
  - stiffness loss / cracks
  - new unmodeled dynamics

---

### Event mapping

- Environmental load anomaly → **T²**
- Structural fracture → **Q statistic**

Reason:
- loads remain within learned correlation structure
- damage introduces new residual directions outside model space



## **Part B: Parametric Factor Partition Engine & UI Architecture**

The objective is to implement an end-to-end data processing and diagnostic visualization pipeline that instantiates a parametric **Factor Analysis (FA)** engine with **Varimax rotation**. The engine must partition empirical observation metrics into a shared structural subspace and a distinct diagonal noise floor, mapping the results onto a highly structured, asymmetric grid interface.

### Develop a Mathematical Engine & Pre-processing Guardrails


Implement an isolated algorithm to structure and decompose empirical data structures according to the following guidelines:

1. **Z-Score Mapping Layer:** Accept an arbitrary data frame and isolate numeric observation columns. Standardize the observations into standard normal space $Z \in \mathbb{R}^{n \times m}$ applying an unbiased sample standard deviation ($ddof=1$). Build an explicit numerical guardrail that catches zero-variance channels ($\sigma_j = 0$) and caps them to a minimum boundary threshold of $10^{-15}$ to prevent divide-by-zero runtime exceptions.
2. **Dimensionality Constraint Layer:** Let $m$ represent the count of isolated sensor features, and $k$ represent the target count of latent physical factors. Program two validation rule exceptions:
* Reject processing if the operational snapshot count ($n$) is less than or equal to the channel count ($m$).
* Throw an descriptive error if the requested latent space dimension ($k$) is not strictly less than the available feature space dimension ($m$).


3. **Information Decomposition Layer:** Fit a Factor Analysis framework utilizing a Maximum Likelihood or Singular Value Decomposition (SVD) algorithm combined with an orthogonal Varimax rotation matrix ($T$). From the fitted parameters, compute the vector arrays for:
* **Communality Vector ($h^2 \in \mathbb{R}^m$):** The total shared variance captured across each channel.
* **Uniqueness Vector ($\varphi^2 \in \mathbb{R}^m$):** The remaining isolated sensor noise floor.

In [9]:
import numpy as np
import pandas as pd
from sklearn.decomposition import FactorAnalysis
from sklearn.preprocessing import StandardScaler
from sklearn.utils.extmath import svd_flip


class FactorAnalysisEngine:
    """
    End-to-end mathematical engine for:
    - Z-score normalization with variance guardrails
    - Dimensionality validation
    - Factor Analysis (SVD-based)
    - Varimax rotation
    - Communality + uniqueness extraction
    """

    def __init__(self, df: pd.DataFrame):
        self.df = df.copy()

    # -----------------------------
    # 1. Z-score mapping layer
    # -----------------------------
    def _zscore_transform(self):
        X = self.df.select_dtypes(include=[np.number]).values.astype(float)

        if X.shape[1] == 0:
            raise ValueError("No numeric columns found in dataset.")

        std = X.std(axis=0, ddof=1)

        # Guardrail: prevent divide-by-zero instability
        std = np.where(std < 1e-15, 1e-15, std)

        mean = X.mean(axis=0)

        Z = (X - mean) / std

        self.Z = Z
        self.m = Z.shape[1]
        self.n = Z.shape[0]

        return Z

    # -----------------------------
    # 2. Dimensionality constraints
    # -----------------------------
    def _validate(self, k):
        if self.n <= self.m:
            raise ValueError(
                f"Invalid regime: n={self.n} must be > m={self.m} for stable factor estimation."
            )

        if k >= self.m:
            raise ValueError(
                f"Invalid latent dimension: k={k} must be strictly less than m={self.m}."
            )

    # -----------------------------
    # 3. Varimax rotation
    # -----------------------------
    def _varimax(self, L, gamma=1.0, q=50, tol=1e-6):
        """
        Orthogonal varimax rotation
        """
        p, k = L.shape
        R = np.eye(k)

        d = 0

        for i in range(q):
            d_old = d
            Lambda = L @ R

            u, s, vh = np.linalg.svd(
                L.T @ (Lambda**3 - (gamma / p) * Lambda @ np.diag(np.diag(Lambda.T @ Lambda)))
            )

            R = u @ vh
            d = np.sum(s)

            if d_old != 0 and d / d_old < 1 + tol:
                break

        return L @ R

    # -----------------------------
    # 4. Fit Factor Model
    # -----------------------------
    def fit(self, k):
        Z = self._zscore_transform()
        self._validate(k)

        # Factor Analysis (MLE / SVD equivalent approximation)
        fa = FactorAnalysis(n_components=k, random_state=0)
        fa.fit(Z)

        loadings = fa.components_.T  # m x k

        # Varimax rotation
        rotated_loadings = self._varimax(loadings)

        self.loadings = rotated_loadings

        # -----------------------------
        # 5. Communality & Uniqueness
        # -----------------------------
        communality = np.sum(rotated_loadings**2, axis=1)

        uniqueness = np.maximum(1 - communality, 1e-15)

        self.communality = communality
        self.uniqueness = uniqueness

        return {
            "Z": Z,
            "loadings": rotated_loadings,
            "communality": communality,
            "uniqueness": uniqueness
        }


# -----------------------------
# Example usage
# -----------------------------
# engine = FactorAnalysisEngine(df)
# result = engine.fit(k=2)
# print(result["communality"])
# print(result["uniqueness"])

### Latent Score Projection

1. **Mathematical State Reconstruction:** Using Thomson’s Minimum Mean Squared Error (MMSE) regression methodology, transform the standard normal snapshot array ($Z$) directly into hidden factor states $F \in \mathbb{R}^{n \times k}$ using the system relation:

$$F = Z R^{-1} \Lambda_{\text{rotated}}$$


2. **Performance Diagnostic Telemetry:** Calculate the empirical variance of the reconstructed factor scores enforcing an unbiased scale correction ($ddof=1$). Print a clean confirmation breakdown out to the logging console that presents the calculated system-wide metrics:

$$\text{Average System Communality \%} = \frac{1}{m}\sum_{j=1}^m h_j^2 \times 100$$


$$\text{Average System Uniqueness \%} = \frac{1}{m}\sum_{j=1}^m \varphi_j^2 \times 100$$

In [10]:

import numpy as np

class FactorScoreProjection:
    """
    Latent factor score reconstruction + system diagnostics
    using Thomson MMSE regression.
    """

    def __init__(self, Z, loadings, communality, uniqueness):
        """
        Z: (n x m) standardized data
        loadings: (m x k) rotated factor loadings Λ_rotated
        communality: (m,)
        uniqueness: (m,)
        """
        self.Z = Z
        self.L = loadings
        self.h2 = communality
        self.psi2 = uniqueness

        self.n, self.m = Z.shape
        self.k = loadings.shape[1]

    # -----------------------------
    # 1. Factor score reconstruction
    # -----------------------------
    def reconstruct_factors(self):
        """
        F = Z R^{-1} Λ_rotated
        where R = ΛΛ^T + Ψ (model-implied covariance)
        """

        # Model-implied covariance of Z
        R = self.L @ self.L.T + np.diag(self.psi2)

        # Numerical stability
        R_inv = np.linalg.pinv(R)

        # Thomson MMSE projection
        F = self.Z @ R_inv @ self.L

        self.F = F
        self.R = R

        return F

    # -----------------------------
    # 2. Diagnostic telemetry
    # -----------------------------
    def diagnostics(self):
        """
        Computes system-level communality & uniqueness metrics
        """

        avg_comm = np.mean(self.h2) * 100
        avg_uni = np.mean(self.psi2) * 100

        # Factor score variance (unbiased)
        if hasattr(self, "F"):
            factor_var = np.var(self.F, axis=0, ddof=1)
        else:
            factor_var = None

        print("\n=== SYSTEM DIAGNOSTICS ===")
        print(f"Average System Communality (%): {avg_comm:.4f}")
        print(f"Average System Uniqueness (%): {avg_uni:.4f}")

        if factor_var is not None:
            print("\nFactor Score Variances (ddof=1):")
            for i, v in enumerate(factor_var):
                print(f"F{i+1}: {v:.6f}")

        return {
            "avg_communality_percent": avg_comm,
            "avg_uniqueness_percent": avg_uni,
            "factor_variance": factor_var
        }


# -----------------------------
# Example pipeline usage
# -----------------------------
# Assume:
# Z, loadings, h2, psi2 already computed from previous module

# projector = FactorScoreProjection(Z, loadings, h2, psi2)
# F = projector.reconstruct_factors()
# metrics = projector.diagnostics()

### Layout Engineering & Interface Design

Render a unified, production-grade **2 × 2 diagnostic canvas** using Plotly Subplots that enforces the following strict geometric layout constraints to ensure maximum readability and prevent overlapping label blocks:

* **Grid Specifications:** Configure the dashboard matrix with an explicit `horizontal_spacing=0.24` and a `vertical_spacing=0.28`. Set the global canvas parameters to `width=1250` and `height=750` utilizing a clean `plotly_white` theme.
* **Legend Formatting:** Place a single, horizontally oriented legend centered directly above the subplots (`orientation="h", x=0.5, y=1.02`). Enforce clean `margin` padding structures to isolate the interface elements (`t=150, b=60, l=140, r=80`).

Populate the grid layout using the four specific tracking profiles below:

```
                  2 × 2 FACTOR DIAGNOSTICS GRID MATRIX
+------------------------------------------+------------------------------------------+
|             [Row 1, Col 1]               |             [Row 1, Col 2]               |
|      Structural Loadings Heatmap         |     Horizontal Stacked Bar Chart         |
|   - Absolute values |λ_(j,r)|            |   - Shared Communality (h²) in Blue      |
|   - Color scale: 'YlOrRd'                |   - Isolated Uniqueness (φ²) in Orange   |
|   - Left-anchored scale bar (x = -0.15)  |   - Total space constraints: 0 to 100%   |
+------------------------------------------+------------------------------------------+
|             [Row 2, Col 1]               |             [Row 2, Col 2]               |
|         Sensor Noise Floor Line          |       Latent Empirical Variance          |
|   - Dot-dashed red line graph with markers|   - Vertical distribution bar chart      |
|   - Tracks exact Uniqueness values (φ²)  |   - Verifies energy level allocations    |
|   - Enforce tick angle adjustments (25°) |   - Black boundary outlines (width = 0.5)|
+------------------------------------------+------------------------------------------+

```

1. **Subplot (1,1): Structural Loadings Matrix Heatmap**
Render the absolute value of the rotated component weights matrix ($|\Lambda_{\text{rotated}}|$) using a continuous `YlOrRd` (Yellow-Orange-Red) color palette. To protect the clarity of the sensor text identifiers along the y-axis, isolate the continuous scale colorbar to the left side of the canvas by explicitly setting `x=-0.15`, `len=0.38`, `y=0.78`, `yanchor="middle"`, and `xanchor="right"`.
2. **Subplot (1,2): Variance Partitioning Profile**
Build a horizontal stacked bar chart mapping out the tracking allocation per sensor channel. The base layer must represent the percentage contribution of shared communalities ($h^2 \times 100$) utilizing color hex code `#1f77b4`. Stack the corresponding sensor uniqueness values ($\varphi^2 \times 100$) directly on top utilizing color hex code `#ff7f0e`. Force the x-axis limits strictly between `[0, 100]`.
3. **Subplot (2,1): Sensor Uniqueness Line Profile**
Construct a clear line chart tracking the isolated channel noise thresholds ($\varphi^2$) across all monitored hardware positions. Apply a red (`#d62728`), 2-pixel wide dot-dashed line string combined with visual `x` indicators. Apply a custom horizontal label layout optimization by rotating the category ticks to a distinct `tickangle=25` constraint.
4. **Subplot (2,2): Latent Factor Variance Distribution**
Draw a vertical bar chart displaying the empirical variance computed across the extracted common factor score vectors. Color the bars green (`#2ca02c`) with an explicit black border envelope (`width=0.5`).

In [11]:
import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots

def plot_factor_diagnostics(Z, loadings, communality, uniqueness, F):
    """
    2x2 Factor Diagnostics Canvas
    """

    m, k = loadings.shape

    sensor_labels = [f"S{j+1}" for j in range(m)]
    factor_labels = [f"F{j+1}" for j in range(k)]

    # -----------------------------
    # Prepare derived quantities
    # -----------------------------
    comm_pct = communality * 100
    uniq_pct = uniqueness * 100

    factor_var = np.var(F, axis=0, ddof=1)

    # -----------------------------
    # Create subplot canvas
    # -----------------------------
    fig = make_subplots(
        rows=2, cols=2,
        subplot_titles=[
            "Structural Loadings Heatmap",
            "Variance Partitioning Profile",
            "Sensor Uniqueness Profile",
            "Latent Empirical Variance"
        ],
        horizontal_spacing=0.24,
        vertical_spacing=0.28
    )

    # -----------------------------
    # (1,1) Heatmap |Λ|
    # -----------------------------
    fig.add_trace(
        go.Heatmap(
            z=np.abs(loadings),
            x=factor_labels,
            y=sensor_labels,
            colorscale="YlOrRd",
            colorbar=dict(
                x=-0.15,
                len=0.38,
                y=0.78,
                yanchor="middle",
                xanchor="right"
            )
        ),
        row=1, col=1
    )

    # -----------------------------
    # (1,2) Stacked bar: communality + uniqueness
    # -----------------------------
    fig.add_trace(
        go.Bar(
            y=sensor_labels,
            x=comm_pct,
            orientation="h",
            name="Communality (%)",
            marker=dict(color="#1f77b4")
        ),
        row=1, col=2
    )

    fig.add_trace(
        go.Bar(
            y=sensor_labels,
            x=uniq_pct,
            orientation="h",
            name="Uniqueness (%)",
            marker=dict(color="#ff7f0e")
        ),
        row=1, col=2
    )

    # -----------------------------
    # (2,1) Uniqueness line plot
    # -----------------------------
    fig.add_trace(
        go.Scatter(
            x=sensor_labels,
            y=uniqueness,
            mode="lines+markers",
            name="Uniqueness φ²",
            line=dict(color="#d62728", width=2, dash="dot"),
            marker=dict(symbol="x", size=10)
        ),
        row=2, col=1
    )

    # -----------------------------
    # (2,2) Factor variance bars
    # -----------------------------
    fig.add_trace(
        go.Bar(
            x=factor_labels,
            y=factor_var,
            name="Factor Variance",
            marker=dict(color="#2ca02c", line=dict(color="black", width=0.5))
        ),
        row=2, col=2
    )

    # -----------------------------
    # Layout styling
    # -----------------------------
    fig.update_layout(
        template="plotly_white",
        width=1250,
        height=750,
        barmode="stack",
        legend=dict(
            orientation="h",
            x=0.5,
            y=1.02,
            xanchor="center"
        ),
        margin=dict(t=150, b=60, l=140, r=80)
    )

    # Axis tuning
    fig.update_xaxes(tickangle=25, row=2, col=1)
    fig.update_yaxes(title_text="Sensors", row=1, col=1)

    return fig


# -----------------------------
# Example usage (expects prior outputs)
# -----------------------------
# fig = plot_factor_diagnostics(Z, loadings, h2, psi2, F)
# fig.show()

# **Subspace Diagnostics & Feature De-correlation**

In structural health monitoring (SHM) and industrial asset diagnostics, multi-sensor grids generate highly continuous, collinear data streams. To establish reliable anomaly detection limits, engineers must project these high-dimensional observation spaces onto lower-dimensional sub-spaces.

This assignment evaluates two fundamental methods for dimensionality reduction using an identical 4-channel sensor array (`Sensor_1` to `Sensor_4`):

1. **Principal Component Analysis (PCA):** A geometric, non-parametric approach aimed at total variance maximization.
2. **Factor Analysis (FA):** A generative, parametric framework aimed at partitioning shared structural variance from unique localized sensor noise.



## **Part A: Develop a Feature Analysis Dashboard**

**UI Layout 1: PCA Optimization Suite ($2\times  3$ Horizontal Layout)**

The first interface must display a 5-panel horizontal grid layout.
* **Panel 1,1 (Heatmap):** Feature Loadings Matrix $|P_{j,r}|$.
* **Panel 1,2 (Bar Chart):** Absolute Eigenvalues ($\lambda_j$) tracking variance magnitude.
* **Panel 1,3 (Combined Bar + Line):** Marginal Explained Variance % per axis (bars) overlaid with Cumulative Captured Variance Curve (dashed line).
* **Panel 2,1 (Bar Chart):** Residual Space (Unexplained Variance %) showing excluded information ($100 - \text{Cumulative Variance \%}$).
* **Panel 2,2 (Scatter Line):** Sample Mean Hotelling's $T^2$ tracking distance vs. Subspace Size $k$.
* **Panel 2,3 (Scatter Line):** Sample Mean $Q$ Statistic (SPE) tracking residual decay vs. Truncation Cutoff $k$.

**UI Layout 2: FA Latent Subspace Suite ($2 \times 2$ Grid Layout)**

The second interface must structure its subplots into a balanced $2 \times 2$ matrix.

* **Panel 1,1 (Heatmap):** Structural Loadings Matrix $|\lambda_{j,r}|$.
* **Panel 1,2 (Stacked Horizontal Bar):** Variance Allocation breakdown showing Shared Communality ($h^2$) versus Channel Uniqueness ($\varphi^2$) totaling $100\%$ per sensor.
* **Panel 2,1 (Scatter Line):** Sensor Uniqueness Noise Floor Profile ($\varphi^2$) across monitored channels to instantly isolate instrument faults.
* **Panel 2,2 (Bar Chart):** Latent Factor Scores Empirical Variance showing the post-rotation power of the extracted physical dimensions.


In [12]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from sklearn.decomposition import FactorAnalysis

# -----------------------------
# Synthetic data
# -----------------------------
np.random.seed(42)
n_samples = 2500

f1 = np.random.normal(0, 1, n_samples)
f2 = np.random.normal(0, 1, n_samples)

s1 = 0.85*f1 + 0.10*f2 + np.random.normal(0, 0.3, n_samples)
s2 = 0.80*f1 + 0.15*f2 + np.random.normal(0, 0.35, n_samples)
s3 = 0.12*f1 + 0.90*f2 + np.random.normal(0, 0.25, n_samples)
s4 = 0.02*f1 + 0.05*f2 + np.random.normal(0, 1.40, n_samples)

df_asset = pd.DataFrame(
    np.vstack([s1, s2, s3, s4]).T,
    columns=["Sensor_1", "Sensor_2", "Sensor_3", "Sensor_4"]
)

# -----------------------------
# Standardization
# -----------------------------
X = df_asset.values
feature_names = df_asset.columns.tolist()

X_mean = X.mean(axis=0)
X_std = X.std(axis=0, ddof=1)
X_std[X_std == 0] = 1e-15

Z = (X - X_mean) / X_std
n, m = Z.shape

# -----------------------------
# PCA engine
# -----------------------------
S = np.cov(Z, rowvar=False, ddof=1)

eigvals, eigvecs = np.linalg.eigh(S)
idx = np.argsort(eigvals)[::-1]

eigvals = eigvals[idx]
eigvecs = eigvecs[:, idx]

scores = Z @ eigvecs

explained_var = eigvals / eigvals.sum()
cumulative_var = np.cumsum(explained_var)
residual_var = 1 - cumulative_var

k_values = np.arange(1, m)

mean_T2 = []
mean_Q = []

for k in k_values:
    scores_k = scores[:, :k]
    lambdas_k = eigvals[:k]
    residual_scores = scores[:, k:]

    T2 = np.sum((scores_k ** 2) / lambdas_k, axis=1)
    Q = np.sum(residual_scores ** 2, axis=1)

    mean_T2.append(np.mean(T2))
    mean_Q.append(np.mean(Q))

# -----------------------------
# PCA dashboard
# -----------------------------
fig_pca = make_subplots(
    rows=2,
    cols=3,
    subplot_titles=(
        "PCA Loadings Matrix |P|",
        "Absolute Eigenvalues",
        "Explained Variance",
        "Residual Space %",
        "Mean Hotelling T² vs k",
        "Mean Q/SPE vs k"
    ),
    horizontal_spacing=0.08,
    vertical_spacing=0.22
)

fig_pca.add_trace(
    go.Heatmap(
        z=np.abs(eigvecs),
        x=[f"PC{i+1}" for i in range(m)],
        y=feature_names,
        colorscale="YlOrRd",
        text=np.round(np.abs(eigvecs), 3),
        texttemplate="%{text}"
    ),
    row=1,
    col=1
)

fig_pca.add_trace(
    go.Bar(
        x=[f"PC{i+1}" for i in range(m)],
        y=eigvals,
        name="Eigenvalues"
    ),
    row=1,
    col=2
)

fig_pca.add_trace(
    go.Bar(
        x=[f"PC{i+1}" for i in range(m)],
        y=explained_var * 100,
        name="Marginal Explained %"
    ),
    row=1,
    col=3
)

fig_pca.add_trace(
    go.Scatter(
        x=[f"PC{i+1}" for i in range(m)],
        y=cumulative_var * 100,
        mode="lines+markers",
        name="Cumulative Explained %",
        line=dict(dash="dash")
    ),
    row=1,
    col=3
)

fig_pca.add_trace(
    go.Bar(
        x=[f"PC{i+1}" for i in range(m)],
        y=residual_var * 100,
        name="Residual %"
    ),
    row=2,
    col=1
)

fig_pca.add_trace(
    go.Scatter(
        x=k_values,
        y=mean_T2,
        mode="lines+markers",
        name="Mean T²"
    ),
    row=2,
    col=2
)

fig_pca.add_trace(
    go.Scatter(
        x=k_values,
        y=mean_Q,
        mode="lines+markers",
        name="Mean Q"
    ),
    row=2,
    col=3
)

fig_pca.update_layout(
    title="PCA Optimization Suite",
    template="plotly_white",
    width=1350,
    height=750,
    showlegend=True
)

fig_pca.show()

# -----------------------------
# Varimax rotation
# -----------------------------
def varimax(Phi, gamma=1.0, q=50, tol=1e-6):
    p, k = Phi.shape
    R = np.eye(k)
    d_old = 0

    for _ in range(q):
        Lambda = Phi @ R
        u, s, vh = np.linalg.svd(
            Phi.T @ (
                Lambda**3
                - (gamma / p) * Lambda @ np.diag(np.diag(Lambda.T @ Lambda))
            )
        )
        R = u @ vh
        d_new = np.sum(s)

        if d_old != 0 and d_new / d_old < 1 + tol:
            break

        d_old = d_new

    return Phi @ R

# -----------------------------
# FA engine
# -----------------------------
k_factors = 2

fa = FactorAnalysis(n_components=k_factors, random_state=42)
fa.fit(Z)

loadings = fa.components_.T
rotated_loadings = varimax(loadings)

communality = np.sum(rotated_loadings**2, axis=1)
uniqueness = 1 - communality
uniqueness = np.clip(uniqueness, 0, 1)

R_corr = np.cov(Z, rowvar=False, ddof=1)
factor_scores = Z @ np.linalg.inv(R_corr) @ rotated_loadings
factor_variance = np.var(factor_scores, axis=0, ddof=1)

factor_names = [f"Factor_{i+1}" for i in range(k_factors)]

# -----------------------------
# FA dashboard
# -----------------------------
fig_fa = make_subplots(
    rows=2,
    cols=2,
    subplot_titles=(
        "FA Structural Loadings Matrix",
        "Variance Allocation: h² vs φ²",
        "Sensor Uniqueness Noise Floor",
        "Latent Factor Score Variance"
    ),
    horizontal_spacing=0.24,
    vertical_spacing=0.28
)

fig_fa.add_trace(
    go.Heatmap(
        z=np.abs(rotated_loadings),
        x=factor_names,
        y=feature_names,
        colorscale="YlOrRd",
        text=np.round(np.abs(rotated_loadings), 3),
        texttemplate="%{text}",
        colorbar=dict(
            title="|Loading|",
            x=-0.15,
            len=0.38,
            y=0.78,
            yanchor="middle",
            xanchor="right"
        )
    ),
    row=1,
    col=1
)

fig_fa.add_trace(
    go.Bar(
        y=feature_names,
        x=communality * 100,
        orientation="h",
        name="Shared Communality h²",
        marker=dict(color="#1f77b4")
    ),
    row=1,
    col=2
)

fig_fa.add_trace(
    go.Bar(
        y=feature_names,
        x=uniqueness * 100,
        orientation="h",
        name="Uniqueness φ²",
        marker=dict(color="#ff7f0e")
    ),
    row=1,
    col=2
)

fig_fa.add_trace(
    go.Scatter(
        x=feature_names,
        y=uniqueness,
        mode="lines+markers",
        name="Uniqueness φ²",
        line=dict(color="#d62728", width=2, dash="dashdot")
    ),
    row=2,
    col=1
)

fig_fa.add_trace(
    go.Bar(
        x=factor_names,
        y=factor_variance,
        name="Latent Factor Variance",
        marker=dict(color="#2ca02c", line=dict(color="black", width=0.5))
    ),
    row=2,
    col=2
)

fig_fa.update_layout(
    title="FA Latent Subspace Suite",
    template="plotly_white",
    width=1250,
    height=750,
    barmode="stack",
    legend=dict(orientation="h", x=0.5, y=1.02, xanchor="center"),
    margin=dict(t=150, b=60, l=140, r=80)
)

fig_fa.update_xaxes(range=[0, 100], row=1, col=2)
fig_fa.update_xaxes(tickangle=25, row=2, col=1)

fig_fa.show()

# -----------------------------
# Summary tables
# -----------------------------
pca_summary = pd.DataFrame({
    "PC": [f"PC{i+1}" for i in range(m)],
    "Eigenvalue": eigvals,
    "Marginal Explained Variance %": explained_var * 100,
    "Cumulative Explained Variance %": cumulative_var * 100,
    "Residual Variance %": residual_var * 100
})

fa_summary = pd.DataFrame({
    "Sensor": feature_names,
    "Communality_h2": communality,
    "Uniqueness_phi2": uniqueness
})

display(pca_summary)
display(fa_summary)

,PC,Eigenvalue,Marginal Explained Variance %,Cumulative Explained Variance %,Residual Variance %
0,PC1,1.974420,49.360512,49.360512,5.063949e+01
1,PC2,1.006968,25.174209,74.534721,2.546528e+01
2,PC3,0.881088,22.027198,96.561920,3.438080e+00
3,PC4,0.137523,3.438080,100.000000,1.110223e-14


,Sensor,Communality_h2,Uniqueness_phi2
0,Sensor_1,0.864586,0.135414
1,Sensor_2,0.875204,0.124796
2,Sensor_3,0.277897,0.722103
3,Sensor_4,0.008658,0.991342


## **Part B: Analysis**

### Generate the following synthetic data and answer the questions below

The synthetic multi-sensor asset data must be generated using
```python
import numpy as np
import pandas as pd

# Set execution seed for perfect reproducible properties
np.random.seed(42)
n_samples = 2500

# 1. Generate underlying, independent true physical driving forces (Latent Factors)
f1 = np.random.normal(0, 1, n_samples)  # Primary structural mode/load
f2 = np.random.normal(0, 1, n_samples)  # Secondary operational mode

# 2. Construct the Observable Multi-Sensor Array with targeted noise profiles
# Sensors 1 and 2 couple strongly to the primary process (f1)
s1 = 0.85 * f1 + 0.10 * f2 + np.random.normal(0, 0.3, n_samples)
s2 = 0.80 * f1 + 0.15 * f2 + np.random.normal(0, 0.35, n_samples)

# Sensor 3 couples strongly to the secondary process (f2)
s3 = 0.12 * f1 + 0.90 * f2 + np.random.normal(0, 0.25, n_samples)

# Sensor 4 suffers an extreme electrical/instrumentation fault (massive localized white noise)
s4 = 0.02 * f1 + 0.05 * f2 + np.random.normal(0, 1.40, n_samples)

# 3. Compile into the final evaluation DataFrame
df_asset = pd.DataFrame(
    data=np.vstack([s1, s2, s3, s4]).T,
    columns=['Sensor_1', 'Sensor_2', 'Sensor_3', 'Sensor_4']
)

```

**Empirical Covariance and Correlation Properties**

* **Sensors 1, 2, and 3:** High cross-correlation due to strong coupling with the common underlying physical structures ($f_1$ and $f_2$).
* **Sensor 4:** Highly un-correlated with all other channels, but exhibits an exceptionally large variance magnitude ($\sigma^2 \approx 2.0$) due to its high noise contamination scale.


### Question 1: The Total Variance Illusion (PCA Subplots 1 & 2 vs. FA Subplots 1 & 2)

* **Context:** In your PCA dashboard, `PC 1` captures roughly $45\%$ of the total system variance, and `PC 2` captures over $30\%$, indicating a highly structured, multi-component system. However, looking at the Factor Analysis dashboard, the **Variance Partitioning Stacked Bar** reveals that `Sensor_4` has a **Uniqueness value ($\varphi^2$) exceeding $98\%$**.
* **Core Problem:** 1. What does a Uniqueness value close to $100\%$ tell you about the physical nature of the data coming from `Sensor_4`?
2. Because PCA maximizes total global variance without distinguishing between types of variance, how does the massive localized noise of `Sensor_4` "trick" the PCA engine into inflating its top eigenvalues?
3. Why is this a major risk if an engineer relies *only* on PCA to build an automated plant anomaly detection framework?



### Question 2: Decoupling Structural Loading via Rotation (FA Subplot 1 vs. PCA Eigenvectors)

* **Context:** Look at the **Structural Loadings Matrix** heatmap in the top-left of the FA dashboard. You can see that `Sensor_1` and `Sensor_2` load heavily on `Factor 1`, while `Sensor_3` loads exclusively on `Factor 2`.
* **Core Problem:**
1. Traditional PCA forces its principal axes to follow a strict mathematical hierarchy ($\lambda_1 > \lambda_2 > \dots$) where every component tries to sweep up as much mixed variance as possible across all sensors. Explain how the **Varimax rotation** in Factor Analysis relaxes this rule to create "simple structure" (clean, isolated groupings).
2. From a plant operator's standpoint, why is the rotated FA heatmap much easier to troubleshoot during a structural failure than a raw, unrotated PCA eigenvector matrix?





### Question 3: Determining Subspace Truncation ($k$) using $T^2$ and $Q$ Profiles

* **Context:** Look at the right side of the PCA dashboard, showing **Mean Hotelling's $T^2$** and **Mean $Q$ Statistic (SPE)** plotted against the truncation cutoff $k$.
* **Core Problem:**
1. Describe the exact behavior of the Mean $Q$ Statistic curve as the subspace cutoff transitions from $k=1$ to $k=2$, and then from $k=2$ to $k=3$.
2. Why does a sharp drop followed by a distinct "flat elbow" in the $Q$ statistic identify **$k=2$** as the true hidden physical dimensionality of this asset?
3. If you incorrectly choose a truncation cutoff of $k=3$ for your monitoring system, what type of data are you forcing into your "clean" principal subspace?

### Question 4: Operational Trade-offs in System Health Monitoring

* **Context:** Imagine you are deploying a real-time predictive maintenance pipeline for a fleet of identical industrial assets based on these exact models.
* **Core Problem:** Compare the operational trade-offs of these two approaches:
1. **The PCA Strategy:** Tracking the main principal subspace using Hotelling's $T^2$ alongside a residual noise floor monitor using the $Q$ statistic.
2. **The FA Strategy:** Monitoring the rotated Latent Factor Scores directly.


* Which strategy is more robust against a single sensor losing calibration or experiencing an electrical short? Justify your choice using the **Sensor Uniqueness Noise Floor ($\varphi^2$)** metrics.

## Question 1: The Total Variance Illusion

### 1. Meaning of Sensor_4 uniqueness close to 100%

From the FA result,

$$
\phi_4^2 = 0.991342
$$

or approximately

$$
99.13\%.
$$

This means Sensor_4 is almost completely dominated by unique sensor-specific noise. Its communality is only

$$
h_4^2 = 0.008658
$$

or approximately

$$
0.87\%.
$$

Therefore, Sensor_4 does not significantly represent the shared physical structure of the asset. It is likely affected by high localized noise, calibration error, or instrumentation fault.

---

### 2. How Sensor_4 tricks PCA

PCA maximizes total variance without knowing whether the variance is useful structural variance or useless sensor noise.

From the PCA result,

$$
PC_1 = 49.36\%
$$

and

$$
PC_2 = 25.17\%.
$$

Together,

$$
PC_1 + PC_2 = 74.53\%.
$$

However, Sensor_4 has a very high loading on PC2:

$$
|P_{Sensor4,PC2}| = 0.95.
$$

This means PC2 is strongly influenced by Sensor_4. Since Sensor_4 is mostly noise according to FA, PCA incorrectly treats this noisy variance as an important principal direction.

---

### 3. Risk of relying only on PCA

If an engineer uses PCA alone, Sensor_4 noise may be treated as real structural information.

This creates major risks:

- False alarms may increase.
- True structural changes may be hidden.
- The PCA eigenvalues may become inflated by noise.
- Monitoring thresholds may be built around faulty sensor behavior.

Therefore, PCA alone can be misleading when one sensor has very high localized noise.

---

## Question 2: Decoupling Structural Loading via Rotation

### 1. How Varimax creates simple structure

PCA follows the eigenvalue hierarchy

$$
\lambda_1 > \lambda_2 > \lambda_3 > \lambda_4.
$$

Each principal component tries to capture maximum total variance. Therefore, PCA components can mix useful structural variance and sensor noise.

In the PCA loading matrix, Sensor_4 strongly loads on PC2:

$$
|P_{Sensor4,PC2}| = 0.95.
$$

This shows that PCA gives high importance to Sensor_4 because of its large variance.

In Factor Analysis, Varimax rotation produces a simpler and more interpretable structure. From the FA loading matrix:

$$
Sensor_1 \rightarrow Factor_1 = 0.930
$$

$$
Sensor_2 \rightarrow Factor_1 = 0.924
$$

$$
Sensor_3 \rightarrow Factor_2 = 0.482
$$

Sensor_1 and Sensor_2 clearly belong to Factor_1. Sensor_3 is more related to Factor_2. Sensor_4 has very weak loadings on both factors.

This is called simple structure.

---

### 2. Why FA heatmap is easier for troubleshooting

The FA heatmap is easier for a plant operator because it separates the sensors according to hidden physical factors.

From the results:

- Sensor_1 and Sensor_2 are strongly connected to Factor_1.
- Sensor_3 is mainly connected to Factor_2.
- Sensor_4 is mostly noise.

Therefore, if Factor_1 changes, the operator can focus on the physical system measured by Sensor_1 and Sensor_2.

If Sensor_4 changes, the operator should suspect sensor noise or sensor malfunction rather than structural damage.

PCA does not provide this clear physical separation.

---

## Question 3: Determining Subspace Truncation Using $T^2$ and $Q$

### 1. Behavior of the Mean $Q$ curve

The residual statistic is

$$
Q_i = \sum_{j=k+1}^{m}z_{i,j}^2.
$$

From the PCA result, the Mean $Q$ curve decreases as $k$ increases.

When

$$
k=1 \rightarrow k=2,
$$

Mean $Q$ drops sharply from about

$$
2.0
$$

to about

$$
1.0.
$$

This means the second principal component captures a large amount of previously unexplained energy.

When

$$
k=2 \rightarrow k=3,
$$

Mean $Q$ drops again to about

$$
0.14.
$$

After this point, only a very small residual remains.

---

### 2. Why the elbow identifies hidden dimensionality

The cumulative explained variance values are:

$$
PC_1 = 49.36\%
$$

$$
PC_1+PC_2 = 74.53\%
$$

$$
PC_1+PC_2+PC_3 = 96.56\%.
$$

Since the residual variance becomes very small after

$$
k=3,
$$

the practical elbow appears at

$$
k=3.
$$

This means the PCA model needs three components to capture almost all system variance.

However, FA shows that one of the components is likely influenced by Sensor_4 noise. Therefore, physically meaningful hidden structure is mainly represented by two latent factors, while the third PCA component mostly captures remaining variance/noise.

---

### 3. What happens if $k=3$ is chosen incorrectly?

If

$$
k=3
$$

is selected as the clean principal subspace, the model includes almost all variance:

$$
96.56\%.
$$

However, because Sensor_4 is noisy, part of this retained variance is not structural.

Thus, choosing

$$
k=3
$$

forces noisy sensor behavior into the clean principal subspace.

This may reduce the sensitivity of anomaly detection because the model begins to accept noise as normal behavior.

---

## Question 4: Operational Trade-offs in System Health Monitoring

### PCA Strategy

The PCA strategy monitors:

$$
T^2
$$

inside the retained principal subspace and

$$
Q
$$

in the residual subspace.

PCA is useful because it compresses data and removes correlation.

However, PCA cannot separate structural variance from sensor noise.

In this result, Sensor_4 has a high PCA loading on PC2:

$$
|P_{Sensor4,PC2}| = 0.95.
$$

So PCA gives high importance to Sensor_4 even though FA shows it is mostly noise.

---

### FA Strategy

The FA strategy models

$$
R \approx \Lambda\Lambda^T+\Psi.
$$

It separates shared structural variance from unique sensor noise.

From the FA result:

$$
\phi_1^2 = 0.135414
$$

$$
\phi_2^2 = 0.124796
$$

$$
\phi_3^2 = 0.722103
$$

$$
\phi_4^2 = 0.991342
$$

Sensor_4 has the highest uniqueness, almost equal to 1.

Therefore, FA clearly identifies Sensor_4 as a faulty or noisy sensor.

